### ISSUE TITLE
[Background blur] How to add a background image for video card

### ISSUE THREAD

**Issue by mrnav90** [aws/amazon-chime-sdk-js#1748]
Hi @davcolle @ltrung 
About background filter, I hope Chime SDK team can support add a background image for video filter.
This background filter can be a `image url` or `image path`.

Github: https://github.com/Volcomix/virtual-background
Like demo: https://volcomix.github.io/virtual-background/

Thank you so much.

**Comment by majo35-6c** [comment]
@mrnav90 
you can combine 2 processor
 - VritualBackgorund
 - Chime Blur
 ```
  const videoTransformDevice = new DefaultVideoTransformDevice(logger, deviceId, [VritualBackgorundProcessor, ChimeBlurProcessor]);
 ```

**Comment by mrnav90** [comment]
@majo35-6c 

Thank you for suggestion, I'll try it. But I think Chime SDK Team make a filter option for blur background is that great.
Many thanks

**Comment by majo35-6c** [comment]
@mrnav90 
example how you can implement custom blur or image as background. 

https://github.com/w-okada/chime-videoprocess-ts

 - chime team use google selfie segmentatiom landscape tflite model with 256x144

**Comment by dathacky** [comment]
@majo35-6c 
In case browser supports blur, I need to modify the `drawImageWithMask` function again, I tried appendChild(this.targetCanvas);, it works, but in my video it doesn't appear

`    let processor: BackgroundBlurProcessor;
    if (await BackgroundBlurProcessorProvided.isSupported()) {
      logger.info('Using browser-provided background blur');
      processor = new BackgroundBlurProcessorProvided(spec, options);
    } else {
      logger.info('Using built-in background blur');
      processor = new BackgroundBlurProcessorBuiltIn(spec, options);
    }`

**Comment by majo35-6c** [comment]
@dathacky

Do you want implement video procesor for virtual background with Image and bind result to videoTile?

**Comment by dathacky** [comment]
@majo35-6c 
I customized that function with this.backgroundMode and this.backgroundImage attribute, with BACKGROUND_MODE.BLUR, the videoTile was blur, but when I used BACKGROUND_MODE.IMAGE, that didn't work, I tried appendCanvas this.targetCanvas to html to check it out, i see my code is already running, but videoTile doesn't exist

**Comment by majo35-6c** [comment]
@dathacky

Try draw only image on targetCanvas.
Can you see your image?

**Comment by dathacky** [comment]
<img width="1440" alt="Screen Shot 2021-11-04 at 10 01 31" src="https://user-images.githubusercontent.com/37299929/140250779-e8939e65-94b1-4d1c-b39d-89cb11d6d04f.png">

i appended it to html once

```
  if (count === 0 && this.backgroundImage) {
    const divImage = document.getElementById('1');
    divImage.appendChild(this.targetCanvas);
    count = 1;
  }```

**Comment by dathacky** [comment]
<img width="1420" alt="Screen Shot 2021-11-04 at 10 04 13" src="https://user-images.githubusercontent.com/37299929/140251018-baf3ce7d-9db2-4187-9b07-cd4c4f25158a.png">
@majo35-6c 
With blur, it work

**Comment by majo35-6c** [comment]
@dathacky 

- create custom processor
```
class VirtualBackgroundProcessor {

mode = null;

 awsProcessorDelegate = null;
 
 // output
  targetCanvas = document.createElement('canvas');
  canvasVideoFrameBuffer = new CanvasVideoFrameBuffer(this.targetCanvas);
 
 canvasBackground= document.createElement('canvas');
 
   setBackgroundImage(path) {
   this.canvasBackground.width = 960; // 1280;
    this.canvasBackground.height = 540; // 720;
    const bg = new Image();
    bg.crossOrigin = 'anonymous';
    bg.src = path;
    bg.onload = () => {
      this.canvasBackground.getContext('2d').clearRect(0, 0, this.canvasBackground.width, this.canvasBackground.height);
      this.canvasBackground
        .getContext('2d')
        .drawImage(bg, 0, 0, this.canvasBackground.width, this.canvasBackground.height);
    };
  }
  
  process = async (buffers) => {
  
  if (buffers.length === 0) {
      return buffers;
    }
    const canvasCamera = buffers[0].asCanvasElement();
    const { width, height } = canvasCamera;
    if (!width || !height) {
      return buffers;
    }

    this.targetCanvas.width = width;
    this.targetCanvas.height = height;

    switch (this.mode) {
    case  AWS_BLUR: {
        if (!this.awsProcessorDelegate) {
          this.awsProcessorDelegate = await BackgroundBlurVideoFrameProcessor.create(null, { blurStrength: 30 });
        }
        return this.awsProcessorDelegate.process(buffers);
      }
    }
    case AWS_IMAGE: {
        this.setBackgroundImage(`${ASSETS_URL}${IMAGES_PATH]}`);
        if (!this.awsProcessorDelegate) {
          this.awsProcessorDelegate = await BackgroundBlurVideoFrameProcessor.create(null, { blurStrength: 30 });
        }
        this.awsProcessorDelegate.sourceHeight = canvasCamera.height;
        this.awsProcessorDelegate.sourceWidth = canvasCamera.width;
        this.awsProcessorDelegate.targetCanvas.width = canvasCamera.width;
        this.awsProcessorDelegate.targetCanvas.height = canvasCamera.height;
        this.awsProcessorDelegate.setBlurPixels();

        let mask = this.awsProcessorDelegate.mask$.value;

        const hscale = this.awsProcessorDelegate.spec.model.input.width / canvasCamera.width;
        const vscale = this.awsProcessorDelegate.spec.model.input.height / canvasCamera.height;
        this.awsProcessorDelegate.scaledCanvas = document.createElement('canvas');
        this.awsProcessorDelegate.scaledCanvas.width = canvasCamera.width * hscale;
        this.awsProcessorDelegate.scaledCanvas.height = canvasCamera.height * vscale;

        const scaledCtx = this.awsProcessorDelegate.scaledCanvas.getContext('2d');
        scaledCtx.save();
        scaledCtx.scale(hscale, vscale);
        scaledCtx.drawImage(canvasCamera, 0, 0);
        scaledCtx.restore();

        const imageData = scaledCtx.getImageData(
          0,
          0,
          this.awsProcessorDelegate.scaledCanvas.width,
          this.awsProcessorDelegate.scaledCanvas.height
        );
        const maskPromise = this.awsProcessorDelegate.mask$.whenNext();
        this.awsProcessorDelegate.worker.postMessage({ msg: 'predict', payload: imageData }, [imageData.data.buffer]);
        mask = await maskPromise;

        const drawImageWithMask = (inputCanvas, mask) => {
          if (!mask) {
            mask = new ImageData(
              this.awsProcessorDelegate.spec.model.input.width,
              this.awsProcessorDelegate.spec.model.input.height
            );
          }

          const scaledCtx = this.awsProcessorDelegate.scaledCanvas.getContext('2d');

          scaledCtx.putImageData(mask, 0, 0);

          const { targetCanvas } = this;
          const { width, height } = targetCanvas;

          const canvasCtx = targetCanvas.getContext('2d');

          // draw the mask
          canvasCtx.save();
          canvasCtx.clearRect(0, 0, width, height);
          canvasCtx.drawImage(this.awsProcessorDelegate.scaledCanvas, 0, 0, width, height);

          // Only overwrite existing pixels.
          canvasCtx.globalCompositeOperation = 'source-in';
          // draw image over mask...
          canvasCtx.drawImage(inputCanvas, 0, 0, width, height);

          // draw under person
          canvasCtx.globalCompositeOperation = 'destination-over';
          // canvasCtx.filter = `blur(${15}px)`;
          canvasCtx.drawImage(this.canvasBackground, 0, 0, targetCanvas.width, targetCanvas.height);
          canvasCtx.restore();
        };

        drawImageWithMask(canvasCamera, mask);

        break;
      }
    case .NONE:
      default: {
        this.convert(canvasCamera);
        break;
      }
  
  buffers[0] = this.canvasVideoFrameBuffer;
  return buffers;
  };
  
    convert = (canvas) => {
    this.targetCanvas.width = canvas.width;
    this.targetCanvas.height = canvas.height;
    const ctx = this.targetCanvas.getContext('2d');
    ctx.drawImage(canvas, 0, 0, this.targetCanvas.width, this.targetCanvas.height);
    return this.targetCanvas;
  };

}
```

**Comment by dathacky** [comment]
@majo35-6c 
Thank you very much for this enthusiastic help, I found the problem, it's because I use the image link on the internet

**Comment by mrnav90** [comment]
@awsvikas Could you tell me how about working process this feature?
Thanks

**Comment by awsvikas** [comment]
@mrnav90 - Thank you for your interest in the feature, we are currently researching on options to make this easier to adopt. Thank you for your patience, I will update this issue once we have details.

**Comment by ltrung** [comment]
We have released background replacement that allow users to use background image (See PR https://github.com/aws/amazon-chime-sdk-js/pull/1920). This feature is available in NPM version 2.25.0 so close this issue now. Feel free to try it out and let us know your feedback.
